xclim/indices/_agro.py

```diff
@@ -882,6 +883,188 @@ def water_budget(
+        during `window_dry_end` days are considered a dry sequence. The precipitations must be lower than `thresh_dry_end`
+        for either every day in the sequence (`method_dry_end == "per_day"`) or for the total (`method_dry_end == "total"`).
+    method_dry_end: {"per_day", "total"}
+        Method used to define a dry sequence associated with `window_dry_end`. The threshold `thresh_dry_end` is either compared
+        to every daily precipitations (`method_dry_end == "per_day"`) or to total precipitations (`method_dry_end == "total"`)
+        in the sequence `window_dry` days.
+    date_min_end: DayOfYearStr
+        First day of year when season can end ("mm-dd").
+    date_max_end: DayOfYearStr
+        Last day of year when season can end ("mm-dd").
+    freq : str
+      Resampling frequency.
+
+    Returns
+    -------
+    rain_season_start: xr.DataArray, [dimensionless]
+    rain_season_end: xr.DataArray, [dimensionless]
+    rain_season_length: xr.DataArray, [time]
+
+    Notes
+    -----
+    The rain season starts at the end of a period of raining (a total precipitation  of `thresh_wet_start` over `window_wet_start` days). This must
+    be directly followed by a period of `window_not_dry_start` days with no dry sequence. The dry sequence is a period of `window_dry_start`
+    days where precipitations are below `thresh_dry_start` (either the total precipitations over the period, or the daily precipitations, depending
+    on `method_dry_start`). The rain season stops when a dry sequence happens (the dry sequence is defined as in the start sequence,
+    but with parameters `window_dry_end`, `thresh_dry_end` and `method_dry_end`). The dates on which the season can start are constrained
+    by `date_min_start`and `date_max_start` (and similarly for the end of the season).
+
+    References
+    ----------
+    :cite:cts:`sivakumar_predicting_1998`
+    """
+    # Unit conversion.
+    pram = rate2amount(pr, out_units="mm")
+    thresh_wet_start = convert_units_to(thresh_wet_start, pram)
+    thresh_dry_start = convert_units_to(thresh_dry_start, pram)
+    thresh_dry_end = convert_units_to(thresh_dry_end, pram)
+
+    last_doy = pram.indexes["time"][-1].strftime("%m-%d")
+
+    # Eliminate negative values.
+    pram = xarray.where(pram < 0, 0, pram)
+    pram.attrs["units"] = "mm"
+
+    # should we flag date_min_end  < date_max_start?
+    def _get_first_run(run_positions, start_date, end_date):
+        run_positions = select_time(run_positions, date_bounds=(start_date, end_date))
+        first_start = run_positions.argmax("time")
+        return xarray.where(
+            first_start != run_positions.argmin("time"), first_start, np.NaN
+        )
+
+    # Find the start of the rain season
+    def _get_first_run_start(pram, window_dry=window_dry_start):
+        pram = select_time(pram, date_bounds=(date_min_start, last_doy))
```

Comment:
Why `last_doy` ? Like you could have an incomplete daily series finishing in july, but still having a few full years.

2 ideas:
- Adapt `select_time` so it can accept incomplete tuples.
- Recompute `last_doy` each time.
```suggestion
        last_doy = pram.indexes["time"][-1].strftime("%m-%d")
        pram = select_time(pram, date_bounds=(date_min_start, last_doy))
```

Reply:
Just a side comment:

>    # Eliminate negative values.
    pram = xarray.where(pram < 0, 0, pram)
    pram.attrs["units"] = "mm"
    
    
    This is legacy from Wf_africa, I should just remove this

Reply:
> Adapt select_time so it can accept incomplete tuples

This would be nice, I will take a look

Reply:
> Like you could have an incomplete daily series finishing in july, but still having a few full years.

This is meant to be used in a resampled computation, so I'm not sure if that concern will be important? In each resampled group, the last_doy could be different

Reply:
Indeed! But the current code takes `last_doy` from the full time series, not the resampled group. (Line 983)